### PR TITLE
resourcemanager: add task manager to support forcing to stop the task.

### LINKED ISSUE #40566: support forcing to stop the task in the pool

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Weizhen Wang <wangweizhen@pingcap.com>

### What problem does this PR solve?

Issue Number: close #40566

Problem Summary:

### What is changed and how it works?

 add task manager to support forcing to stop the task.

why do we need to use a task manager to implement it? it is easy to implement the scheduler the subjob in the task in the future.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: resourcemanager/pooltask/task_manager.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+		}
+	}
+	return TaskManager[T, U, C, CT, TF]{
+		task:        task,
+		concurrency: c,
+	}
+}
+
+// RegisterTask register a task to the manager.
+func (t *TaskManager[T, U, C, CT, TF]) RegisterTask(task uint64, concurrency int32) {
```
Comment (Defined2014) [inline resourcemanager/pooltask/task_manager.go:83]: rename `task` to `id` is better.

File: resourcemanager/pooltask/task_manager.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	"container/list"
+	"sync"
+	"time"
+
+	"go.uber.org/atomic"
+)
+
+const shard int = 8
+
+func getShardID(id uint64) uint64 {
```
Comment (Defined2014) [inline resourcemanager/pooltask/task_manager.go:27]: Not get point. Why need shard here?

Comment (hawkingrei) [reply resourcemanager/pooltask/task_manager.go:27]: In the future, we need to schedule the task with the task manager while we will add or delete the task in the task manager from the user business. Lock contention is a bit intense so we add the shard lock.

File: resourcemanager/pooltask/task_manager.go (hunk 1 above)
Comment (hawkingrei) [reply resourcemanager/pooltask/task_manager.go]: Done

File: resourcemanager/pooltask/task_manager.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+type tContainer[T any, U any, C any, CT any, TF Context[CT]] struct {
+	task *TaskBox[T, U, C, CT, TF]
+}
+
+type meta struct {
+	stats    *list.List
+	createTs time.Time
+	origin   int32
+	running  atomic.Int32
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply resourcemanager/pooltask/task_manager.go:39]: We can compare ```running``` with ```pool.Capacity``` to decide the scheduler. but In this PR, I don't add the task scheduler.

[the remark this answers is not in the document]
Comment (hawkingrei) [reply resourcemanager/pooltask/task_manager.go]: Now it is ```int32```. In the next pr, It will be changed.

**Review by Defined2014 — APPROVED** [review]

**Comment by hawkingrei** [comment]
@Benjamin2037 @zimulala PTAL

File: resourcemanager/pooltask/task_manager.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+// Running return the count of running job in this task.
+func (t *TaskManager[T, U, C, CT, TF]) Running(taskID uint64) int32 {
+	shardID := getShardID(taskID)
+	t.task[shardID].rw.Lock()
+	defer t.task[shardID].rw.Unlock()
+	return t.task[shardID].stats[taskID].running
+}
+
+// StopTask is to stop a task by TaskID.
+func (t *TaskManager[T, U, C, CT, TF]) StopTask(taskID uint64) {
```
Comment (zimulala) [inline resourcemanager/pooltask/task_manager.go:129]: Please add UT for `DeleteTask` and `StopTask`

File: util/gpool/spmc/spmcpool.go [hunk 5]
```diff
@@ -426,3 +432,20 @@ func (p *Pool[T, U, C, CT, TF]) revertWorker(worker *goWorker[T, U, C, CT, TF])
...
 	return true
 }
+
+// DeleteTask is to delete task.
+// Please don't use it manually.
+func (p *Pool[T, U, C, CT, TF]) DeleteTask(id uint64) {
+	p.taskManager.DeleteTask(id)
+}
+
+// StopTask is to stop task by id
```
Comment (zimulala) [inline util/gpool/spmc/spmcpool.go:442]: s/id/ID

File: resourcemanager/pooltask/task_manager.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+type meta struct {
+	stats    *list.List
+	createTs time.Time
+	origin   int32
+	running  int32
+}
+
+func newStats(concurrency int32) *meta {
+	s := &meta{
+		createTs: time.Now(),
```
Comment (zimulala) [inline resourcemanager/pooltask/task_manager.go:44]: s/createTs/createTS ?

File: resourcemanager/pooltask/task_manager.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	t.task[shardID].rw.Lock()
+	t.task[shardID].stats[taskID].stats.PushBack(tc)
+	t.task[shardID].stats[taskID].running++ // running job in this task
+	t.task[shardID].rw.Unlock()
+}
+
+// ExitSubTask is to exit a task, and it will decrease the count of running pooltask.
+func (t *TaskManager[T, U, C, CT, TF]) ExitSubTask(taskID uint64) {
+	shardID := getShardID(taskID)
+	t.running.Dec() // total running task
```
Comment (zimulala) [inline resourcemanager/pooltask/task_manager.go:114]:
```suggestion
	t.running.Dec() // total running tasks
```

File: util/gpool/spmc/worker.go [hunk 8]
```diff
@@ -58,14 +58,23 @@ func (w *goWorker[T, U, C, CT, TF]) run() {
...
+					if f.GetStatus() == pooltask.StopTask {
+						f.Done()
+						continue
+					}
 					f.GetResultCh() <- w.pool.consumerFunc(t.Task, f.ConstArgs(), ctx)
 					f.Done()
 				}
+				w.pool.ExitSubTask(f.TaskID())
 			}
+			f.Finish()
```
Comment (zimulala) [inline util/gpool/spmc/worker.go:77]: Could we remove this function or remove `Done`? It seems that `f.Done` and `f.Finish()` do the same thing: `t.wg.Done()`

Comment (hawkingrei) [reply util/gpool/spmc/worker.go:77]: In fact, They are the same thing. But if Finish is also called, here are two ```Done```. it is weird. ```Finish``` is to tell system this ```TaskBox``` has finished job.

File: resourcemanager/pooltask/task.go [hunk 9]
```diff
@@ -39,10 +49,24 @@ type TaskBox[T any, U any, C any, CT any, TF Context[CT]] struct {
...
+}
+
+// SetStatus is to set the status of task.
+func (t *TaskBox[T, U, C, CT, TF]) SetStatus(s int32) {
+	t.status.Store(s)
 }
 
 // NewTaskBox is to create a task box for pool.
 func NewTaskBox[T any, U any, C any, CT any, TF Context[CT]](constArgs C, contextFunc TF, wg *sync.WaitGroup, taskCh chan Task[T], resultCh chan U, taskID uint64) TaskBox[T, U, C, CT, TF] {
+	// we still need to do some work after a taskbox finish.
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply resourcemanager/pooltask/task.go]: Done

File: resourcemanager/pooltask/task.go [hunk 10]
```diff
@@ -31,6 +32,15 @@ func (NilContext) GetContext() any {
 	return nil
 }
 
+const (
+	// PendingTask is a task waiting to start
+	PendingTask int32 = iota
+	// RunningTask is a task running
+	RunningTask
+	// StopTask is a stop task
```
[the remark this answers is not in the document]
Comment (hawkingrei) [reply resourcemanager/pooltask/task.go]: Done

File: resourcemanager/pooltask/task_manager.go (hunk 6 above)
Comment (hawkingrei) [reply resourcemanager/pooltask/task_manager.go:44]: Done

File: resourcemanager/pooltask/task_manager.go (hunk 4 above)
Comment (hawkingrei) [reply resourcemanager/pooltask/task_manager.go:138]: ```DeleteTask``` has been added in ```TestTaskManager``` which is at the ```resourcemanager/pooltask/task_test.go```. ```StopTask``` is in the ```TestStopPool``` which is at the ```util/gpool/spmc/spmcpool_test.go```

File: util/gpool/spmc/worker.go (hunk 8 above)
Comment (hawkingrei) [reply util/gpool/spmc/worker.go:77]: > Could we remove this function or remove `Done`?

If we remove it, we will block when to exit.

**Review by YangKeao — APPROVED** [review]
LGTM

**Review by zimulala — APPROVED** [review]

**Pull request merged**